com.unity.sg2/Editor/GraphUI/DataModel/ShaderGraphModel.cs

```diff
@@ -385,119 +430,13 @@ public static IPortModel FindOutputPortByName(INodeModel nodeModel, string portI
-                    foreach (var node in sources)
-                    {
-                        m_TempNodeWave.Push(node);
-                        m_TempAddedToNodeWave.Add(node);
-                    }
-
-                    while (m_TempNodeWave.Count > 0)
-                    {
-                        var node = m_TempNodeWave.Pop();
-                        if (node == null)
-                            continue;
-
-                        result.Add(node);
-
-                        // grab connected nodes in propagation direction, add them to the node wave
-                        ForeachConnectedNode(node, dir, AddNextLevelNodesToWave);
-                    }
-
-                    // clean up any temp data
-                    m_TempNodeWave.Clear();
-                    m_TempAddedToNodeWave.Clear();
-                }
-        }
-
-        static IEnumerable<NodeHandler> GetUpstreamNodes(NodeHandler startingNode)
-        {
-            return Utils.GraphTraversalUtils.GetUpstreamNodes(startingNode);
-        }
-
-        static IEnumerable<NodeHandler> GetDownstreamNodes(NodeHandler startingNode)
-        {
-            return Utils.GraphTraversalUtils.GetDownstreamNodes(startingNode);
-        }
-
-        static GraphDataNodeModel TryGetNodeModel(ShaderGraphModel shaderGraphModel, NodeHandler inputNodeReader)
-        {
-            // TODO: Make a mapping between every node model and node reader so we can also do lookup
-            // from NodeReaders to NodeModels, as is needed below for instance
-            return null;
-        }
-
-        static void ForeachConnectedNode(GraphDataNodeModel sourceNode, PropagationDirection dir, Action<GraphDataNodeModel> action)
-        {
-            sourceNode.TryGetNodeHandler(out var nodeReader);
-
-            ShaderGraphModel shaderGraphModel = sourceNode.GraphModel as ShaderGraphModel;
-
-            // Enumerate through all nodes that the node feeds into and add them to the list of nodes to inspect
-            if (dir == PropagationDirection.Downstream)
-            {
-                foreach (var connectedNode in GetDownstreamNodes(nodeReader))
-                {
-                    action(TryGetNodeModel(shaderGraphModel, connectedNode));
-                }
-            }
-            else
-            {
-                foreach (var connectedNode in GetUpstreamNodes(nodeReader))
-                {
-                    action(TryGetNodeModel(shaderGraphModel, connectedNode));
-                }
-            }
-        }
-
-        public IEnumerable<GraphDataNodeModel> GetNodesInHierarchyFromSources(IEnumerable<GraphDataNodeModel> nodeSources, PropagationDirection propagationDirection)
+        // TODO: (Sai) Would it be better to have a way to gather any variable nodes
+        // linked to a blackboard item at a GraphHandler level instead of here?
+        public IEnumerable<INodeModel> GetLinkedVariableNodes(string variableName)
```

Comment:
arguably this function should just return the list of graph data names directly by talking to GraphHandler instead of going through GTF models as it currently does

We have the dictionary in GraphStorage that contains the linked reference node mappings, so we could just get that directly, but didn't want to rely on that without running that code by Liz first as we might toss it for a better solution soon

Reply:
Oh-- yea. You could just grab the referable port and then check its output connections (in graph handler), all of which would be owning reference nodes / variable nodes.